jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/HttpReceiverOverHTTP2.java

```diff
@@ -29,49 +28,89 @@
 import org.eclipse.jetty.client.api.Response;
 import org.eclipse.jetty.http.HttpField;
 import org.eclipse.jetty.http.HttpFields;
-import org.eclipse.jetty.http.HttpStatus;
 import org.eclipse.jetty.http.MetaData;
 import org.eclipse.jetty.http2.api.Stream;
+import org.eclipse.jetty.http2.frames.DataFrame;
 import org.eclipse.jetty.http2.frames.HeadersFrame;
 import org.eclipse.jetty.http2.frames.PushPromiseFrame;
 import org.eclipse.jetty.http2.frames.ResetFrame;
 import org.eclipse.jetty.http2.internal.ErrorCode;
 import org.eclipse.jetty.http2.internal.HTTP2Channel;
 import org.eclipse.jetty.http2.internal.HTTP2Stream;
+import org.eclipse.jetty.io.Content;
 import org.eclipse.jetty.io.EndPoint;
 import org.eclipse.jetty.util.Callback;
 import org.eclipse.jetty.util.Promise;
-import org.eclipse.jetty.util.thread.Invocable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class HttpReceiverOverHTTP2 extends HttpReceiver implements HTTP2Channel.Client
 {
     private static final Logger LOG = LoggerFactory.getLogger(HttpReceiverOverHTTP2.class);
 
+    private boolean responseSucceeded;
+
     public HttpReceiverOverHTTP2(HttpChannel channel)
     {
         super(channel);
     }
 
     @Override
-    protected HttpChannelOverHTTP2 getHttpChannel()
+    protected void onInterim()
     {
-        return (HttpChannelOverHTTP2)super.getHttpChannel();
     }
 
     @Override
-    protected void receive()
+    protected void reset()
+    {
+        super.reset();
+        responseSucceeded = false;
+    }
+
+    @Override
+    public Content.Chunk read(boolean fillInterestIfNeeded)
     {
-        // Called when the application resumes demand of content.
         if (LOG.isDebugEnabled())
-            LOG.debug("Resuming response processing on {}", this);
+            LOG.debug("Reading, fillInterestIfNeeded={} in {}", fillInterestIfNeeded, this);
+        Stream stream = getHttpChannel().getStream();
+        Stream.Data data = stream.readData();
+        if (LOG.isDebugEnabled())
+            LOG.debug("Read stream data {} in {}", data, this);
+        if (data == null)
+        {
+            if (fillInterestIfNeeded)
+                stream.demand();
+            return null;
+        }
+        DataFrame frame = data.frame();
+        boolean last = frame.remaining() == 0 && frame.isEndStream();
+        if (last && !responseSucceeded)
+        {
+            responseSucceeded = true;
```

Comment:
Is field `responseSucceeded` necessary? We should be protected by reads after EOF by the base class.

Reply:
No, it's not needed. Removed it.